src/hardware/input/mouseif_dos_driver.cpp

```diff
@@ -920,8 +889,120 @@ void MOUSEDOS_AfterNewVideoMode(const bool setmode)
+
+		if (rows < threshold_lo || rows > threshold_hi) {
+			rows = default_rows;
+		}
+		if (columns < threshold_lo || columns > threshold_hi) {
+			columns = default_columns;
+		}
+
+		state.maxpos_x = static_cast<int16_t>(8 * columns - 1);
+		state.maxpos_y = static_cast<int16_t>(8 * rows - 1);
+	};
+
+	// Set some common/default parameters
+
+	state.granularity_x = 0xffff;
+	state.granularity_y = 0xffff;
+
+	state.minpos_x = 0;
+	state.minpos_y = 0;
+
+	// Apply settings depending on video mode
+
+	switch (bios_screen_mode) {
+	case 0x00: // text, 40x25, black/white        (CGA, EGA, MCGA, VGA)
+	case 0x01: // text, 40x25, 16 colors          (CGA, EGA, MCGA, VGA)
+		state.granularity_x = 0xfff0;
+		state.granularity_y = 0xfff8;
+		set_maxpos_text();
+		// Apply correction due to different x axis granularity
+		state.maxpos_x = static_cast<int16_t>(state.maxpos_x * 2 + 1);
+		break;
+	case 0x02: // text, 80x25, 16 shades of gray  (CGA, EGA, MCGA, VGA)
+	case 0x03: // text, 80x25, 16 colors          (CGA, EGA, MCGA, VGA)
+	case 0x07: // text, 80x25, monochrome         (MDA, HERC, EGA, VGA)
+		state.granularity_x = 0xfff8;
+		state.granularity_y = 0xfff8;
+		set_maxpos_text();
+		break;
+	case 0x08: // 160x200, 16 colors    (PCjr)
+		state.maxpos_x = 159;
+		state.maxpos_y = 199;
+		break;
+	case 0x04: // 320x200, 4 colors     (CGA, EGA, MCGA, VGA)
+	case 0x05: // 320x200, 4 colors     (CGA, EGA, MCGA, VGA)
+	case 0x09: // 320x200, 16 colors    (PCjr)
+		state.maxpos_x = 319;
+		state.maxpos_y = 199;
+		break;
+	case 0x0d: // 320x200, 16 colors    (EGA, VGA)
+	case 0x13: // 320x200, 256 colors   (MCGA, VGA)
+		state.granularity_x = 0xfffe;
+		state.maxpos_x = 639;
+		state.maxpos_y = 199;
+		break;
+	case 0x06: // 640x200, black/white  (CGA, EGA, MCGA, VGA)
+	case 0x0a: // 640x200, 4 colors     (PCjr)
+	case 0x0e: // 640x200, 16 colors    (EGA, VGA)
+		state.maxpos_x = 639;
+		state.maxpos_y = 199;
+		break;
+	case 0x0f: // 640x350, monochrome   (EGA, VGA)
+	case 0x10: // 640x350, 16 colors    (EGA 128K, VGA)
+	           // 640x350, 4 colors     (EGA 64K)
+		state.maxpos_x = 639;
+		state.maxpos_y = 349;
+		break;
+	case 0x11: // 640x480, black/white  (MCGA, VGA)
+	case 0x12: // 640x480, 16 colors    (VGA)
+		state.maxpos_x = 639;
+		state.maxpos_y = 479;
+		break;
+	default: // other modes, most likely SVGA
+		if (!is_svga_mode) {
+			// Unsupported mode, this should probably never happen
+			LOG_WARNING("MOUSE (DOS): Unknown video mode 0x%02x",
+			            bios_screen_mode);
+			// Try to set some sane parameters, do not draw cursor
+			state.inhibit_draw = true;
+			state.maxpos_x = 639;
+			state.maxpos_y = 479;
+		} else if (is_svga_text) {
+			// SVGA text mode
+			state.granularity_x = 0xfff8;
+			state.granularity_y = 0xfff8;
+			set_maxpos_text();
+		} else {
+			// SVGA graphic mode
+			state.maxpos_x = static_cast<int16_t>(CurMode->swidth - 1);
+			state.maxpos_y = static_cast<int16_t>(CurMode->sheight - 1);
+		}
+		break;
```

Comment:
So just checking, how would this work with programs that set up a BIOS video mode first, then tweak it further via the CRTC registers to come up with some custom resolution? I'm guessing if DOSBox also sets the `swidth` and `sheight` params to reflect the tweaked dimensions, this will work perfectly. I'll have to double check that, I guess I'm just wondering if you've tested this at all.

Reply:
> if DOSBox also sets the `swidth` and `sheight` params to reflect the tweaked dimensions

Those come from the video mode 'info block' data structure. They're fixed and represent the official mode 'screen' width and height, among other settings, like the text-mode width and height (`tw` and `th`). They can and can be queried in the VESA 2.x case by the DOS program:

![2023-06-14_16-56](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/507ef94c-ed0c-41e0-be07-d032b238ce3b)

The tweak timing values fall out of vga_draw and then hit the `RENDER_` and `GFX_` functions, and I think there's a call in SDL main that updates the `MOUSE_` side when the mode changes (maybe it could be passed back through there?). 

I'm not sure if the tweaked mode resolution is stuffed back into a VGA register values that we can query (unfortunately).

Reply:
I guess if this newly introduced behaviour is not _worse_ than what we had before, then it's all good and can maybe progressively refine it as needed.

I just wanna make sure we're not introducing regressions for tweaked modes by making decisions based on `swidth` and `sheight`. Like you said, probably this is fine for the defaulting, though, then tweaked modes somehow propagate the changes screen dimensions to the mouse code... "Somehow" being the keyword here, I don't know how it works, neither have the time/inclination now to find out 😄

So I guess it falls to @FeralChild64 to give this some minimal testing in a few tweaked-mode VGA apps to make sure it still works as before.

Reply:
The mouse boundaries are working in Scorched Earth's various tweaked modes.

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/e65b6b56-91ad-42ca-afef-82e6d459c5ab

Reply:
@kcgen

> If these 8s related to 8-dot-mode vs 9-dot-mode text height, this bit is held in the `vga.seq.clocking_mode.is_eight_dot_mode `register.

Unlikely. I have checked that seamless mouse integration works properly in the text mode regardless if the character is 8 or 9 pixels wide. And setting it to 9 would affect the position we report to guest application. I don't think real mouse driver were that smart.